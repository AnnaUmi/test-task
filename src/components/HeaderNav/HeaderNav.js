import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import { HeaderContainer, HeaderLogo, NavMenu, NavItem } from './HeaderNav.styles';

const renderNav = (nav, currRoute) =>
  nav.map(({ route, id, label }) => (
    <Link href={route} key={id}>
      <NavItem isActive={currRoute === route}>
        <div>{label}</div>
      </NavItem>
    </Link>
  ));

const HeaderNav = ({ currMainRoute, routes }) => {
  const logo = routes[0];
  const nav = routes.slice(1, routes.length);

  return (
    <HeaderContainer>
      <Link href={logo.route}>
        <HeaderLogo>{logo.label}</HeaderLogo>
      </Link>
      <NavMenu>{renderNav(nav, currMainRoute)}</NavMenu>
    </HeaderContainer>
  );
};

HeaderNav.propTypes = {
  currMainRoute: PropTypes.string.isRequired,
  routes: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default HeaderNav;
