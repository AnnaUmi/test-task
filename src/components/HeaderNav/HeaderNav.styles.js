import styled from 'styled-components';

export const HeaderContainer = styled.div`
  display: flex;
  background-color: #d3d3d3;
  padding: 0 20px;
  height: 60px;
  align-items: center;
`;

export const HeaderLogo = styled.div`
  border: 1px dashed #696969;
  padding: 5px 0;
  min-width: 120px;
  text-align: center;
  cursor: pointer;
`;

export const NavMenu = styled.div`
  margin-left: 40px;
  display: flex;
`;

export const NavItem = styled.div`
  margin-left: 40px;
  cursor: pointer;
  display: flex;
  flex-direction: column;

  &:first-child {
    margin-left: 0;
  }

  :after {
    content: '';
    display: inline-block;
    width: 100%;
    height: 2px;
    margin-top: 2px;
    background-color: ${({ isActive }) => (isActive ? 'blue' : 'rgba(0,0,0,0)')};
  }

  :hover {
    :after {
      content: '';
      display: inline-block;
      width: 100%;
      height: 2px;
      visibility: 'visible';
      background-color: ${({ isActive }) => (isActive ? 'blue' : '#696969')};
    }
  }
`;
