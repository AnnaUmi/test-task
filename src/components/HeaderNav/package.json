{
  "name": "header-nav",
  "version": "0.0.0",
  "private": true,
  "main": "./HeaderNav.js"
}
