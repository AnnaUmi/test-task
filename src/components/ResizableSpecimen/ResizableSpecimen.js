import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { unstable_renderSubtreeIntoContainer as renderSubtreeIntoContainer } from 'react-dom';
import raf from 'raf';
import Resizable from 're-resizable';
import reactElementToString from '../HighlightedCode/reactElementToString';
import HighlightedCode from '../HighlightedCode';

import {
  ResizableSlider,
  ResizableContainer,
  resizeableStyleProps,
  iFrameStyles,
  Toggle,
  Container,
} from './ResizableSpecimen.styled';

const Slider = () => <ResizableSlider />;

const resizeEnableProps = {
  top: false,
  right: true,
  bottom: false,
  left: false,
  topRight: false,
  bottomRight: false,
  bottomLeft: false,
  topLeft: false,
};

export default class ResizableSpecimen extends Component {
  static resizeEnableProps = resizeEnableProps;

  static slider = Slider;

  static propTypes = {
    children: PropTypes.node.isRequired,
    responsive: PropTypes.bool,
  };

  static defaultProps = {
    responsive: false,
  };

  state = {
    iframeHeight: 0,
    showCodeSnippet: false,
  };

  componentDidMount() {
    this.renderFrameContents();
    this.updateMaResizableWidth();
  }

  updateMaResizableWidth = () => {
    const maxResizableWidth = this.container.clientWidth;
    this.setState({ maxResizableWidth });
  };

  updateHeight = () => {
    const doc = this.iframe.contentDocument;
    const iframeHeight = doc.body.firstChild.offsetHeight;
    this.setState({ iframeHeight });
  };

  toggleCodeSnippet = () => {
    const { showCodeSnippet } = this.state;
    this.setState({
      showCodeSnippet: !showCodeSnippet,
    });
  };

  renderFrameContents = () => {
    if (!this.iframe) {
      return;
    }

    const { children } = this.props;
    const doc = this.iframe.contentDocument;

    if (doc && doc.readyState === 'complete') {
      const contents = <div>{children}</div>;

      // React warns when you render directly into the body since browser
      // extensions also inject into the body and can mess up React.
      doc.body.innerHTML = '<div></div>';
      doc.head.innerHTML = '<style>html,body{margin:0;padding:0;}<style>';

      const base = doc.createElement('base');
      base.setAttribute('href', window.location.href);
      doc.head.appendChild(base);

      // Clone styles from parent document head into the iframe
      const pageStyles = Array.from(document.querySelectorAll('head > style'));

      pageStyles.forEach(s => {
        const clonedNode = s.cloneNode(true);

        // Take styles inserted with insertRule and turn into regular styles inside a style tag
        // cloneNode does not clone styles inserted with insertRule
        // This fix allows to render style components inside an iframe in production
        const insertedStyles = s.sheet.cssRules;
        const insertedStylesArray = Array.prototype.slice.call(insertedStyles);
        insertedStylesArray.forEach(cssRule => {
          clonedNode.textContent += cssRule.cssText;
        });

        doc.head.appendChild(clonedNode);
      });

      renderSubtreeIntoContainer(this, contents, doc.body.firstChild, () => {
        const onRender = content => {
          this.contentHeight = content.offsetHeight;
          this.setState({ iframeHeight: this.contentHeight });
        };

        raf(() => {
          onRender(doc.body.firstChild);
        });
      });
    } else {
      setTimeout(this.renderFrameContents, 0);
    }
  };

  render() {
    const { children, responsive } = this.props;
    const { iframeHeight, maxResizableWidth, showCodeSnippet } = this.state;
    const code = reactElementToString(children);

    return (
      <Container>
        <Toggle onClick={this.toggleCodeSnippet}>{'<>'}</Toggle>
        <ResizableContainer
          innerRef={el => {
            this.container = el;
          }}
        >
          <Resizable
            enable={ResizableSpecimen.resizeEnableProps}
            style={resizeableStyleProps}
            minWidth="320"
            maxWidth={maxResizableWidth}
            handleComponent={{ right: ResizableSpecimen.slider }}
            onResize={() => responsive && this.updateHeight()}
          >
            {responsive && (
              <iframe
                title="demo"
                style={{ ...iFrameStyles, height: `${iframeHeight}px` }}
                ref={ref => {
                  this.iframe = ref;
                }}
              />
            )}
            {/* to get styled components styles injected to the page */}
            <div style={{ display: responsive ? 'none' : 'block' }}>{children}</div>
          </Resizable>
        </ResizableContainer>
        <HighlightedCode code={code} isVisible={showCodeSnippet} />
      </Container>
    );
  }
}
