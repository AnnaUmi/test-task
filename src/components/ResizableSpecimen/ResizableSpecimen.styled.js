import styled from 'styled-components';

const SLIDER_WIDTH = 32;
export const SLIDER_OFFSET = SLIDER_WIDTH - 5;

export const ResizableContainer = styled.div`
  width: calc(100% - ${SLIDER_OFFSET}px);
`;

export const ResizableSlider = styled.div`
  height: 100%;
  flex: 0 0 auto;
  justify-content: center;
  align-items: center;
  width: ${SLIDER_WIDTH}px;
  border-top-right-radius: 6px;
  border-bottom-right-radius: 6px;
  background: #fff;
  box-shadow: 0 12px 36px rgba(0, 0, 0, 0.2);
  color: #5c6ac4;
  cursor: ew-resize;
`;

export const iFrameStyles = {
  width: '100%',
  border: 'none',
};

export const resizeableStyleProps = {
  boxSizing: 'content-box',
  backgroundColor: '#f4f5fa',
  borderRadius: '6px',
  padding: '20px',
  paddingRight: '25px',
};

export const Toggle = styled.span`
  position: absolute;
  right: 5px;
  color: #999;
  top: -25px;
  font-weight: 600;
  cursor: pointer;
`;

export const Container = styled.div`
  position: relative;
`;
