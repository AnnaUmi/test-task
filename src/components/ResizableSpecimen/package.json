{
  "name": "resiable-specimen",
  "version": "0.0.0",
  "private": true,
  "main": "./ResizableSpecimen.js"
}
