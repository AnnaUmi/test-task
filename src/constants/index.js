import PropTypes from 'prop-types';

/* eslint-disable import/prefer-default-export */
export const NORMALIZE_URL =
  'https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css';

export const SIDE_NAV_ITEM_TYPES = {
  ListItem: 'ListItem',
  SubMenu: 'SubMenu',
};

export const routerProps = PropTypes.shape({
  pathName: PropTypes.string,
  route: PropTypes.string,
  query: PropTypes.object,
  back: PropTypes.func,
  beforePopState: PropTypes.func,
  prefetch: PropTypes.func,
  push: PropTypes.func,
  reload: PropTypes.func,
  replace: PropTypes.func,
});

export const sideNavSubMenuItemProps = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  url: PropTypes.string,
};

export const sideNavItemProps = {
  ...sideNavSubMenuItemProps,
  type: PropTypes.oneOf([SIDE_NAV_ITEM_TYPES.ListItem, SIDE_NAV_ITEM_TYPES.SubMenu]).isRequired,
};

export const noop = () => {};
