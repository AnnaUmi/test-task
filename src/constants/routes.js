import { SIDE_NAV_ITEM_TYPES } from '.';

export const HEADER_NAV_ROUTES = [
  {
    route: '/',
    label: 'PRISM Logo',
    id: '/',
  },
  {
    route: '/brand',
    label: 'Brand',
    id: '/brand',
  },
  {
    route: '/product',
    label: 'Product',
    id: '/product',
  },
  {
    route: '/marketing',
    label: 'Marketing',
    id: '/marketing',
  },
  {
    route: '/components',
    label: 'Components',
    id: '/components',
  },
];

const componentsSideNav = [
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Buttons',
    id: 'components/buttons',
    url: '/components/buttons',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Typography',
    id: 'components/typography',
    url: '/components/typography',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Grid',
    id: 'components/grid',
    url: '/components/grid',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Spacing',
    id: 'components/spacing',
    url: '/components/spacing',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Colors',
    id: 'components/colors',
    url: '/components/colors',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Breadcrumbs',
    id: 'components/breadcrumbs',
    url: '/components/breadcrumbs',
  },
];

export const SIDE_NAV_ROUTES = {
  '/components': componentsSideNav,
};
