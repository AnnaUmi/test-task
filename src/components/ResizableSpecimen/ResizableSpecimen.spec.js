import React from 'react';
import 'jest-styled-components';
import { render } from 'react-testing-library';

import ResizableSpecimen from './ResizableSpecimen';
import HeaderNav from '../HeaderNav';
import { HEADER_NAV_ROUTES } from '../../constants/routes';

describe('ResizableSpecimen', () => {
  it('should match snapshot when responsive', () => {
    const { container } = render(
      <ResizableSpecimen responsive>
        <div>test</div>
      </ResizableSpecimen>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when NOT responsive', () => {
    const { container } = render(
      <ResizableSpecimen>
        <div>test</div>
      </ResizableSpecimen>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when responsive and has nested react element', () => {
    const { container } = render(
      <ResizableSpecimen responsive>
        <HeaderNav routes={HEADER_NAV_ROUTES} currMainRoute="/components" />
      </ResizableSpecimen>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
