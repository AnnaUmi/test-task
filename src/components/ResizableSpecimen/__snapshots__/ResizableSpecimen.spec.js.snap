// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableSpecimen should match snapshot when NOT responsive 1`] = `
<div
  class="sc-bZQynM iMWVsx"
>
  <span
    class="sc-EHOje bGDYXH"
  >
    &lt;&gt;
  </span>
  <div
    class="sc-bxivhb dhoNcC"
  >
    <div
      style="position: relative; box-sizing: border-box; background-color: rgb(244, 245, 250); border-radius: 6px; padding: 20px 25px 20px 20px; width: auto; height: auto; min-width: 320; max-width: 0;"
    >
      <div
        style="display: block;"
      >
        <div>
          test
        </div>
      </div>
      <span>
        <div
          style="position: absolute; width: 10px; height: 100%; top: 0px; right: -5px; cursor: col-resize;"
        >
          <div
            class="sc-ifAKCX iygRPX"
          />
        </div>
      </span>
    </div>
    <div
      class="__resizable_base__"
      style="width: 100%; height: 100%; position: absolute; transform: scale(0, 0); left: -2147483647px;"
    />
  </div>
</div>
`;

exports[`ResizableSpecimen should match snapshot when responsive 1`] = `
<div
  class="sc-bZQynM iMWVsx"
>
  <span
    class="sc-EHOje bGDYXH"
  >
    &lt;&gt;
  </span>
  <div
    class="sc-bxivhb dhoNcC"
  >
    <div
      style="position: relative; box-sizing: border-box; background-color: rgb(244, 245, 250); border-radius: 6px; padding: 20px 25px 20px 20px; width: auto; height: auto; min-width: 320; max-width: 0;"
    >
      <iframe
        style="width: 100%; height: 0px;"
        title="demo"
      />
      <div
        style="display: none;"
      >
        <div>
          test
        </div>
      </div>
      <span>
        <div
          style="position: absolute; width: 10px; height: 100%; top: 0px; right: -5px; cursor: col-resize;"
        >
          <div
            class="sc-ifAKCX iygRPX"
          />
        </div>
      </span>
    </div>
    <div
      class="__resizable_base__"
      style="width: 100%; height: 100%; position: absolute; transform: scale(0, 0); left: -2147483647px;"
    />
  </div>
</div>
`;

exports[`ResizableSpecimen should match snapshot when responsive and has nested react element 1`] = `
<div
  class="sc-bZQynM iMWVsx"
>
  <span
    class="sc-EHOje bGDYXH"
  >
    &lt;&gt;
  </span>
  <div
    class="sc-bxivhb dhoNcC"
  >
    <div
      style="position: relative; box-sizing: border-box; background-color: rgb(244, 245, 250); border-radius: 6px; padding: 20px 25px 20px 20px; width: auto; height: auto; min-width: 320; max-width: 0;"
    >
      <iframe
        style="width: 100%; height: 0px;"
        title="demo"
      />
      <div
        style="display: none;"
      >
        <div
          class="sc-gzVnrw qErhy"
        >
          <div
            class="sc-htoDjs bKYpwJ"
          >
            PRISM Logo
          </div>
          <div
            class="sc-dnqmqq dHCrQG"
          >
            <div
              class="sc-iwsKbI eFLAho"
            >
              <div>
                Brand
              </div>
            </div>
            <div
              class="sc-iwsKbI eFLAho"
            >
              <div>
                Product
              </div>
            </div>
            <div
              class="sc-iwsKbI eFLAho"
            >
              <div>
                Marketing
              </div>
            </div>
            <div
              class="sc-iwsKbI fkIfaL"
            >
              <div>
                Components
              </div>
            </div>
          </div>
        </div>
      </div>
      <span>
        <div
          style="position: absolute; width: 10px; height: 100%; top: 0px; right: -5px; cursor: col-resize;"
        >
          <div
            class="sc-ifAKCX iygRPX"
          />
        </div>
      </span>
    </div>
    <div
      class="__resizable_base__"
      style="width: 100%; height: 100%; position: absolute; transform: scale(0, 0); left: -2147483647px;"
    />
  </div>
</div>
`;
